src/typed-request.ts

```diff
@@ -1,21 +1,96 @@
 /** Type-safe wrapper around fetch() for REST APIs */
 
 import {compile} from 'path-to-regexp';
-import {HTTPVerb} from './api-spec';
+import {Endpoint, GetEndpoint, HTTPVerb} from './api-spec';
 
-import {ExtractRouteParams, SafeKey, DeepReadonly, PathsForMethod} from './utils';
+import {
+  ExtractRouteParams,
+  SafeKey,
+  DeepReadonly,
+  PathsForMethod,
+  ValueIntersection,
+  SimplifyType,
+} from './utils';
 
-type ExtractRouteParamsVarArgs<T extends string> = {} extends ExtractRouteParams<T>
-  ? []
-  : [params: Readonly<ExtractRouteParams<T>>];
+type PlaceholderEmpty = null | {[pathParam: string]: never};
+
+// No query or path params -> don't take any arguments
+// No query params -> only take path params (one argument)
+// No path params ->
+//   Either null or {} is permitted as the path argument.
+//   Query param is mandatory if there are non-optional query params
+type ParamVarArgs<Params, Query> = DeepReadonly<
+  [Query] extends [never]
+    ? // This must have arisen from an impossible intersection of query types.
+      [
+        error: '❌ Must specify an HTTP method (get, post, etc.) to apiUrlMaker for this endpoint',
+        _: never,
+      ]
+    : [Query] extends [null]
+    ? [{}] extends [Params]
+      ? [] // no path params, no query
+      : [params: Params] // path params, no query allowed
+    : [{}, {}] extends [Params, Query]
+    ? // No path params, optional query params; zero arg call is OK
+      [params?: PlaceholderEmpty, query?: Query]
+    : [
+        params: [{}] extends Params ? PlaceholderEmpty : Params,
+        ...query: [{}] extends [Query] ? [query?: Query] : [query: Query]
+      ]
+>;
+
+type QueryTypes<P, Methods extends keyof P> = {
+  [M in Methods]: SafeKey<P[M], 'query'>;
+};
+
+// Like Pick<T, K>, but doesn't require that K extends keyof T
+type LoosePick<T, K> = Pick<T, K & keyof T>;
+
+// This tries to determine which query parameters are required to generate a URL.
+// This is easy for a single method, but somewhat involved if we don't know the method.
+// In that case, we try to calculate the intersection of the query parameter types for all
+// methods as if they were "closed" aka "exact" types.
+type SafeQueryTypesForMethod<
+  Endpoint,
+  M extends keyof Endpoint,
+  Q extends QueryTypes<Endpoint, M> = QueryTypes<Endpoint, M>,
+  V extends ValueIntersection<Q> = ValueIntersection<Q>,
+  P extends LoosePick<V, keyof Q[M]> = LoosePick<V, keyof Q[M]>
+> = [V] extends [never | null]
```

Comment:
That is correct. This is all removed over in #17 since the approach no longer needs an intersection.

re: the tuple, it's to prevent distribution for the conditional type. In particular `never` behaves oddly when used in conditional types (`X extends number ? 'yes' : 'no'` will evaluate to `never` if `X` is `never`, rather than `'no'`).